lib/rules/require-camelcase-or-uppercase-identifiers.js

```diff
@@ -79,10 +88,161 @@
  * var snake_case = { SnakeCase: 6 };
  * ```
  *
+ * ##### Valid for `{ allowedPrefix: ["opt_", /pfx\d+_/] }`
+ * ```js
+ * var camelCase = 0;
+ * var CamelCase = 1;
+ * var _camelCase = 2;
+ * var camelCase_ = 3;
+ * var UPPER_CASE = 4;
+ * var opt_camelCase = 5;
+ * var pfx32_camelCase = 6;
+ * ```
+ *
+ * ##### Invalid for `{ allowedPrefix: ["opt_", /pfx\d+/] }`
+ * ```js
+ * var lower_case = 1;
+ * var Mixed_case = 2;
+ * var mixed_Case = 3;
+ * var req_camelCase = 4;
+ * var pfx_CamelCase = 5;
+ * ```
+ *
+ * ##### Valid for `{ allowedSuffixes: ["_dCel", {regex:{pattern:"_[kMG]?Hz"}}] }`
+ * ```js
+ * var camelCase = 0;
+ * var CamelCase = 1;
+ * var _camelCase = 2;
+ * var camelCase_ = 3;
+ * var UPPER_CASE = 4;
+ * var camelCase_dCel = 5;
+ * var _camelCase_MHz = 6;
+ * ```
+ *
+ * ##### Invalid for `{ allowedSuffixes: ["_dCel", {regex:{pattern:"_[kMG]?Hz"}}] }`
+ * ```js
+ * var lower_case = 1;
+ * var Mixed_case = 2;
+ * var mixed_Case = 3;
+ * var camelCase_cCel = 4;
+ * var CamelCase_THz = 5;
+ * ```
+ *
+ * ##### Valid for `{ allExcept: ["var_args", {regex:{pattern:"^ignore",flags:"i"}}] }`
+ * ```js
+ * var camelCase = 0;
+ * var CamelCase = 1;
+ * var _camelCase = 2;
+ * var camelCase_ = 3;
+ * var UPPER_CASE = 4;
+ * var var_args = 5;
+ * var ignoreThis_Please = 6;
+ * var iGnOrEeThis_Too = 7;
+ * ```
+ *
+ * ##### Invalid for `{ allExcept: ["var_args", {regex:{pattern:"^ignore",flags:"i"}}] }`
+ * ```js
+ * var lower_case = 1;
+ * var Mixed_case = 2;
+ * var mixed_Case = 3;
+ * var var_arg = 4;
+ * var signore_per_favore = 5;
+ * ```
  */
 
 var assert = require('assert');
 
+// Convert an array of String or RegExp or ESTree RegExpLiteral values
+// into an array of String or RegExp values.  Returns undefined if the
+// input does not match expectations.
+function processArrayOfStringOrRegExp(iv) {
+    if (!Array.isArray(iv)) {
+        return;
+    }
+    var rv = [];
+    var i = 0;
+    while (rv && (i < iv.length)) {
+        var elt = iv[i];
+        if (typeof elt === 'string') {
+            // string values OK
+            rv.push(elt);
+        } else if (elt instanceof RegExp) {
+            // existing RegExp OK
+            rv.push(elt);
+        } else if (elt && (typeof elt === 'object')) {
+            try {
+                // ESTree RegExpLiteral ok if it produces RegExp
+                rv.push(new RegExp(elt.regex.pattern, elt.regex.flags || ''));
+            } catch (e) {
+                // Not a valid RegExpLiteral
+                rv = null;
+            }
+        } else {
+            // Unknown value
+            rv = null;
```

Comment:
An `assert` (and corresponding exception) is probably better than these `rv = null` silent failures.